### PR TITLE
Login

### PR DESCRIPTION
**Pull request by Restioson**
Checklist:
- [x] Basic login
- [x] Editing username, display name and password
  - [x] Implement on client
- [x] Authentication tokens
  - [x] Check token is valid before allowing use
  - [x] Clean up expired tokens periodically
  - [x] "Save password" feature on client via saving tokens
  - [x] Allow token to expire if not used for a certain duration
  - [x] Token becomes stale if not used for a certain duration and must be refreshed manually with password
  - [x] Check permissions of token before allowing actions
    - [x] Test
- [x] Extra user data (e.g profile picture)
  - [x] GDPR/POPIA/etc. considerations
- [x] Normalize and lowercase username
- [x] Compromise security (e.g what to do after compromise) 
  - [x] Check user is not locked/compromised before allowing login
  - [x] Test
- [x] Encrypt traffic between client and server
  - [x] Integrate into Docker
- [ ] Audit (cc @xacrimon)
  - [x] Preliminary
  - [ ] Final (before merge)

### CODE REVIEW

File: client/gtk/src/main.rs [hunk 1]
```diff
@@ -59,18 +59,22 @@ impl Update for Win {
     type Msg = VertexMsg;
 
     fn model(relm: &Relm<Win>, args: VertexArgs) -> VertexModel {
+        println!(
+            "Connecting to {}",
+            args.ip.clone().unwrap_or("127.0.0.1:8080".to_string())
```
Comment (gegy1000) [inline client/gtk/src/main.rs]: Should extract this to avoid the duplication (and cloning)

File: server/src/database.rs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+impl DatabaseServer {
+    fn create_tables(&mut self) -> impl Future<Item = (), Error = ()> {
+        self.pool
+            .connection()
+            .and_then(|mut conn| {
+                conn.client
+                    .prepare(
+                        "CREATE TABLE IF NOT EXISTS users (
```
Comment (gegy1000) [inline server/src/database.rs]: Maybe should make these their own files and `include_str!`

Comment (Restioson) [reply server/src/database.rs]: Maybe indeed

File: client/gtk/src/main.rs (hunk 1 above)
Comment (Restioson) [reply client/gtk/src/main.rs]: Wdm?

Comment (gegy1000) [reply client/gtk/src/main.rs]: `unwrap_or` is done on the ip twice. Could be extracted into its own variable

**Comment by Restioson** [comment]
Do we need to mitigate homoglyphs in usernames?

File: server/src/database.rs (hunk 2 above)
Comment (Restioson) [reply server/src/database.rs]: May also be clearer inline

**Comment by Restioson** [comment]
What extra user data do we need to store? @gegy1000

**Comment by Restioson** [comment]
And I wonder what the implications of the GDPR and POPIA (the South African equivalent act) are for data storage, and what we need to therefore implement.

**Comment by Restioson** [comment]
So far I haven't implemented any high level tools for post-compromise situations, but I have added a `compromised` boolean and a version of password hash/scheme in the database for users and login tokens.

**Comment by Restioson** [comment]
I've committed a bit more and am ticking off features according to local branch. Unfortunately I can't push yet because my Internet is down.

**Comment by Restioson** [comment]
I don't think so far we have incurred any special GDPR/POPIA requirements. Though, when we implement admin tools, then we will need to have some kind of way to delete users and their data.

With regards to extra user data, I've just added the capability for the server to serve profile pictures at the route `/images/profile_pictures/{id}`, served from within the `./files` directory. Later on we can flesh this out more.

File: client/backend/src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,6 @@
+// TODO save password feature through saving tokens
```
Comment (gegy1000) [inline client/backend/src/lib.rs]: Can remove this TODO now

File: client/backend/src/lib.rs [hunk 4]
```diff
@@ -120,9 +134,18 @@ impl Vertex {
         }
     }
 
-    pub fn login(&mut self) -> Result<(), Error> {
+    pub fn create_user(
+        &mut self,
+        username: &str,
+        display_name: &str,
```
Comment (gegy1000) [inline client/backend/src/lib.rs]: I guess we should consider whether we should do global display name or rather have display names per server like discord?

File: server/src/auth.rs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+fn valid_username(username: &str, config: &Config) -> bool {
+    username.len() <= config.max_username_len as usize
+        && username.len() >= config.min_username_len as usize
+}
+
+pub struct TooShort;
+
+pub fn process_username(username: &str, _config: &Config) -> String {
```
Comment (gegy1000) [inline server/src/auth.rs]: `normalize_username`?

File: server/src/client/server.rs [hunk 6]
```diff
@@ -1,100 +1,110 @@
...
+                .for_each(|s| s.do_send(LogoutThisSession));
         }
     }
 
-    fn send_to_room(&mut self, room: &RoomId, message: ServerMessage, sender: &SessionId) {
+    fn logout_sessions(&mut self, logout: LogoutSessions) {
+        // TODO could probably be optimised
+        for (user_id, device_id) in logout.list {
+            if let Some(online_devices) = self.online_devices.get(&user_id) {
+                let device_id = online_devices.iter().find(|id| **id == device_id);
```
Comment (gegy1000) [inline server/src/client/server.rs]: I think this is just a `.contains`?

File: server/src/auth.rs (hunk 5 above)
Comment (Restioson) [reply server/src/auth.rs]: Normalization is a unicode thing. Basically it makes the string into a consistent and comparable form. We use the most destructive form of normalization, where an anarchist circle-A will get converted to an A. The reason that we do this is to allow logins from different computers with perhaps different input languages (see what happened to Spotify), and also to make it more difficult to impersonate others using homoglyph attacks (e.g cyrillic a vs latin a).

File: server/src/client/server.rs (hunk 6 above)
Comment (Restioson) [reply server/src/client/server.rs]: No, this returns the device id. Contains only checks if it exists.

File: server/src/client/session.rs [hunk 7]
```diff
@@ -208,3 +104,663 @@ impl Handler<SendMessage<ServerMessage>> for ClientWsSession {
...
+
+        self.respond(fut, request_id, ctx)
+    }
+
+    fn verify_user_id_password(
+        &mut self,
+        user_id: UserId,
+        password: String,
+    ) -> impl Future<Item = Result<(), ServerError>, Error = MailboxError> {
+        self.database_server
```
Comment (gegy1000) [inline server/src/client/session.rs]: Interaction with the database server could maybe be extracted to a function for requests specifically. Like it'd map the database errors to internal server error.

File: server/src/config.rs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    #[serde(default = "tokens_sweep_interval_secs")]
+    pub tokens_sweep_interval_secs: u64,
+    #[serde(default = "token_stale_days")]
+    pub token_stale_days: u16,
+    #[serde(default = "token_expiry_days")]
+    pub token_expiry_days: u16,
+}
+
+fn max_password_len() -> u16 {
+    1000
```
Comment (gegy1000) [inline server/src/config.rs]: not a review but hells yeah

File: server/src/config.rs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+fn tokens_sweep_interval_secs() -> u64 {
+    1800 // 30min
+}
+
+fn token_stale_days() -> u16 {
+    7 // 1 week
+}
+
+fn token_expiry_days() -> u16 {
```
Comment (gegy1000) [inline server/src/config.rs]: how are these functions used?

File: server/src/client/session.rs (hunk 7 above)
Comment (Restioson) [reply server/src/client/session.rs]: Huh? What do you mean?

File: server/src/config.rs (hunk 8 above)
Comment (Restioson) [reply server/src/config.rs]: This is just a limit to stop people from being able to DOS us by sending too many characters and expecting us to hash it, lol. All passwords are the same size after hashing.

File: client/backend/src/lib.rs (hunk 4 above)
Comment (Restioson) [reply client/backend/src/lib.rs]: We can have both!

File: server/src/auth.rs (hunk 5 above)
Comment (gegy1000) [reply server/src/auth.rs]: I was proposing to rename it. :P 
But thanks for the explanation! I understand it better now :slightly_smiling_face:

File: server/src/client/server.rs (hunk 6 above)
Comment (gegy1000) [reply server/src/client/server.rs]: You already have the device id

File: server/src/client/session.rs (hunk 7 above)
Comment (gegy1000) [reply server/src/client/session.rs]: There's a lot of repetition in these similar functions that map the error from database to an internal server error type -- I'm proposing a function that would make a request to the database via send but map the error too such that it doesn't have to be repeated

File: server/src/config.rs (hunk 8 above)
Comment (gegy1000) [reply server/src/config.rs]: very epic

File: server/src/auth.rs (hunk 5 above)
Comment (Restioson) [reply server/src/auth.rs]: Ah, I see. `process_username` also makes it lowercase, so it isn't just normalisation, though.

File: server/src/client/server.rs (hunk 6 above)
Comment (Restioson) [reply server/src/client/server.rs]: Oh lol

**Comment by Restioson** [comment]
@gegy1000 any further review comments?

**Comment by gegy1000** [comment]
Nope, all looks great to me! 🙂

**Comment by Restioson** [comment]
Since @xacrimon is busy, perhaps we should merge this into a development branch and then wait for a review to merge that into master?

**Comment by Gegy** [comment]
Sounds good 👍

**Pull request merged**